### ISSUE TITLE
Lock left button single/double click function

### ISSUE THREAD

**Issue by Ar2rZ** [OptiKey/OptiKey#744]
**Is your feature request related to a problem? Please describe.**
Got a request from one of users, who is using single/double click mouse click for a work.

**Describe the solution you'd like**
Single/Double click function lock - so, user locks single/double click button action- so user do not need to click on this button again. Mouse single/double  click is performed till user unlocks single/double click function..

**Describe alternatives you've considered**

**Additional context**

**Comment by SCEhardt** [comment]
@JuliusSweetland  @AdamRoden @kmcnaught 

Any ideas on the best way to make this happen?  Other buttons can be locked because staying on them causes them to press twice in a row.  However on these mouse click buttons, the first press transforms the UI to a mouse click, which the OptiKey button doesn't respond to.  (Then when the mouse click is done, you're back at the initial state.)  

My first thought is maybe there's a way to detect the mouse click on OptiKey and lock the button if it's the same one currently doing the click.  Thoughts on this or other solutions?

**Comment by kmcnaught** [comment]
There was a lot of previous discussion here: https://github.com/OptiKey/OptiKey/issues/421 

It mainly came down to making a decision on a reasonable UX pattern that wouldn't have any ambiguity between whether a user intended to dwell again on the key to lock/unlock it, or instead perform the current mouse event on the Optikey keyboard itself. And how that would behave in edge cases, where e.g. you want to click near the key. 

One option would be to go back to a previous suggestion from @JuliusSweetland of having a "lock mouse actions" key. We could probably extend this from the original suggestion, and have the following UX:
- select "mouse move and left click"
  - performs single click
- select "lock mouse action key" to hold down
- select "mouse move and left click"
  - enters a looping state until either a different mouse action is selected (which would itself be looped), or "lock mouse action" key is released

This would be able to support locking for any mouse actions, e.g. left click, left double click, right click, hover.

**Comment by SCEhardt** [comment]
@AdamRoden I'm not sure where to apply the suggested code; I tried a couple things but didn't see a difference.

@kmcnaught Thanks for linking the other discussion.  I'm still hung up on the problem that mouse clicks don't activate the virtual keyboard.  I think this would lead to a dead-end in either the traditional lock behavior or the new button you suggested:
- If we make the key lockable in the same way as other keys such as Shift: It will never be possible to get to the locked state because the click action doesn't apply to the keyboard, so the button will always reset before it can be selected twice.
- If we make a new lock button for mouse buttons, it will never be possible to escape the locked clicking sequence, because clicks don't affect the keyboard.

So I'm wondering if there's a way to make clicks affect certain keys-- the left-click can click the left-click key; the middle-click can click the middle-click key, and so on.  Then they could lock on the second click and unlock on the third click.

If the above approach seems reasonable: Is there already any click capturing/mapping in Optikey?  Or any tips on where to start research?  I have quite a bit of experience in C# generally but not in the UI that Optikey uses.

**Comment by kmcnaught** [comment]
@SCEhardt you're right - I had forgotten that key selection was disabled whilst a mouse action was underway. I would guess it would be simpler to approach it from the opposite direction - modify the logic so that key selection is still running, but only allowed to affect certain keys - that would avoid needing to worry about how the Optikey app captures mouse clicks. 

Either way you approach it, `MainViewModel.PointToKeyValueMap` is almost certainly going to be helpful

**Comment by kmcnaught** [comment]
and if you want to have a play around, take a look in `HandleFunctionKeySelectionResult` in  `MainViewModel.ServiceEventHandlers.cs`, starting with `case FunctionKeys.MouseMoveAndLeftClick:` [(direct link)](https://github.com/OptiKey/OptiKey/blob/c81b2a252afae9902b1214f556a4432f531cd3e3/src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs#L1546)

you should also look at `SelectionModes` which controls whether a key or point is being selected - having looked it up now I am less confident on my previous suggestion - currently the input sources and progress subscriptions are quite tightly separated as **either** point **or** key selection - it would need a bit of thought into how best to allow a hybrid selection type in either direction.

**Comment by AdamRoden** [comment]
@SCEhardt If you use a dynamic keyboard the xml snippet I offered provides an immediate solution without a massive coding effort.

**Comment by AdamRoden** [comment]
@kmcnaught @SCEhardt Good news, I ran into a situation where I wanted this feature and it motivated me to finally figure out how to make this happen. Look for this bug to be fixed in the next release.

**Comment by JuliusSweetland** [comment]
@kmcnaught This is another feature that is often requested. Can we discuss?

**Comment by AdamRoden** [comment]
@JuliusSweetland @kmcnaught 
Pardon me for speaking out of turn, but I have given this some thought, and I'd like to lay out how one could approach it from a dynamic keyboard starting point.
```
Introduce a new type of KeyCommand capable of triggering different actions based on the fixation timespan
<DynamicKey>
    <Switch>
        <Time Value="500">
            <Action>MouseLeftClick</Action>
        </Time
        <Time Value="1000">
            <Loop>
                <Action>MouseLeftClick</Action>
            </Loop>
        </Time>
    </Switch>
</DynamicKey>

Modify KeyFixationSource to check for one of two scenarios under which it should fire an event for a key:
1. CompletionTime (current behavior)
2. CompletionTime && FixationTimeout (new behavior to be added)

Modifythe KeyFixationSource TriggerSource and Sequence so that when progress >= 1 they send the timespan rather than "1"

Modify InputService SelectionProgressSubscription and SelectionProgress event to pass along the timespan

Modify MainViewModel inputServiceSelectionResultHandler to receive the timespan
Modify MainViewModel.ServiceEventHandler CommandKeyTask to take the timespan into account when processing the KeyCommands for execution

This same logic could be introduced to a built in key by making it utilize the KeyValue.KeyCommand

**Comment by JuliusSweetland** [comment]
Hi Adam - you're never speaking out of turn - this is an open source
project! Besides, you've made some massive contributions...

I like this idea, but can I clarify; what is the timespan that is being
passed as a result of the selection? And why is it important to the
CommandKeyTask?

**Comment by AdamRoden** [comment]
@JuliusSweetland Normally, we pass the KeyValue, but in the case of a "SwitchKey" where one of two sets of actions are possible we have the dilemma of needing to provide the KeyStateService with the original KeyValue of the complete SwitchKey, but the KeyboardOutputService with the derived KeyValue of one of the action sets in the SwitchKey. I was thinking the total timespan of gaze fixation on the key could be passed to determine the derived KeyValue. Maybe it would make more sense to expand the event Tuples and send two KeyValues instead. 

It was just my initial thoughts linking the start and end points.

One thing I didn't mention was how the animation would work, which I figured would be by using the switch times as reset points for the ProgressTriggerSignal, (Kind of an override version of the CompletionTimes code.)

**Comment by JuliusSweetland** [comment]
I'll read this again in the next few days and see if it makes more sense.
I'm struggling to follow this right now.

**Comment by AdamRoden** [comment]
@JuliusSweetland Fresh idea...
Forget everything I said before. Use a new dictionary in the KeyStateService to keep track of the switched state of a key. The user control Key should be able to subscribe to changes and update the UI, (text or symbol).
The KeySelection and KeyProgress events would both interact with KeyStateService to determine whether the key is changing states or actually triggering an action. We may not even need to change the code in KeyFixationSource, which would be nice. I need to explore this further, but it seems promising.

**Comment by JuliusSweetland** [comment]
That sounds much more promising. Nice work.

**Comment by AdamRoden** [comment]
Using KeyStateService does work. I built a crude version where a key will toggle every 500ms of fixation from nothing at 0, a single-click at 500, a double-click at 1000, and back to nothing at 1500. The action only triggers after you release the key. It is reminiscent of shift, ctrl, etc. I didn't go as making the text and symbol change, but could continue on it if you want. Otherwise, is someone else slated to work on this?

**Comment by kmcnaught** [comment]
@AdamRoden I was planning to set aside some time to catch up on this. I haven't understood everything you've described but I'll take a look at the code this week. Let me do that before you put any more work in, and then we can discuss where we're at.

Could you please share a link to a PR or commit for me to try out.

**Comment by AdamRoden** [comment]
@kmcnaught Here is a start: 00fdd907363f7b07c1817cb29fd686afc7aeda41
Here is a sample key:
~~~

<KeyGroup Name="All" LockOnTime="20" CompletionTimes="500,1000"/>
<Content>
  <DynamicKey Col="0" Width="10" Height="2">
    <Switch Symbol="MouseRightClickIcon"><Action>MouseMoveAndRightClick</Action></Switch>
    <Switch Symbol="MouseLeftClickIcon"><Action>MouseMoveAndLeftClick</Action></Switch>
    <Label>Left Click</Label><Symbol>MouseLeftClickIcon</Symbol>
  </DynamicKey>

**Comment by kmcnaught** [comment]
@AdamRoden I'm still refreshing my memory of all the different options and issues we've discussed. I understand what you've done now - this is a clever way to allow the key's state to be set _before_ embarking on a mouse action. I don't understand how this approach would allow you to unset the key's state _after_ you've performed your (continuous) mouse actions - if you were using this to keep clicking, then you wouldn't be able to select the key to turn it off. Did you have a plan towards this goal?

By the way, if I use a key with an infinite loop, I get a StackOverflow recursion when the page is loaded (using https://github.com/OptiKey/OptiKey/commit/00fdd907363f7b07c1817cb29fd686afc7aeda41)

```
<DynamicKey>
	<Label>Space Loop</Label>
	<Symbol>SpaceIcon</Symbol>
	<Loop Count="0">
		<Text>&#32;</Text>
	</Loop>
</DynamicKey>
```

**Comment by kmcnaught** [comment]
@JuliusSweetland I have now written up some notes to remind myself of where we got to in these discussions

### Requirements:

1. The ability to perform repeat left clicks in OK Core
2. The ability to perform repeat left double clicks in OK Core
3. Some way to perform arbitrary repeat mouse events with dynamic keys
4. Preserve the ability to right click on Optikey and left click on the context menu

### Main issue:

Currently the SelectionMode is mutually exclusive point vs key source, which means:
  - After selecting to perform a click action you are no longer able to select a key, e.g. to lock a key down
  - When performing a continuous click you would need to be able to select a key to signal that you want it to end

### User Experience options

1. A setting that makes [left single/double] clicks continuous or not (always)
    - Very limiting, a user can’t easily swap between strategies 
    - Doesn’t win us anything wrt the issues above
    - Wouldn’t require any UI/UX changes in the keyboards 
    
2. Make some mouse keys “lockable” - at least left click and left double click
    - We would need to solve the SelectionMode issue to allow the key to be locked in the first place. One possible implementation is Adam’s “Switch” approach (currently implemented for dynamic keyboards).
    - … and to allow the key to be switched off. This would require simultaneous point and key source functionality (not solved by Adam’s approach unless I've misunderstood something)
    
3. Add a separate key for “Lock mouse actions”. UX would be that a user toggles the state of the lock key, then chooses their mouse action (left click, left double click, etc). 
    - This would allow “continuous left click” to be enabled without needing to worry about the SelectionMode issue
    - We would still need to fix the SelectionMode issue to enable users to get out of the continuous clicking state.
    - We’d have to decide some UX details - does the lock persist until toggled off, or just for one set of continuous actions?     - Which mouse actions does it affect? Is this clear to a user? E.g. does it affect right click? Middle click?

All these options share one key problem - we need a key source running to allow for getting out of the continuous clicking state. When we solve this, the additional issue of locking down a key in the first place is likely solved too. 

Given that all these approaches have the same issue, my vote would be for option (2) in terms of UX, at least for the core Optikey functionality (maybe dynamic keyboards want extra functionality - I haven't thought about that)

So, we need a key selection source and a point selection source to be used in parallel. 

Question 1 - Does this cause any issues with any of our points sources? 

Question 2 - What "key selection" behaviour would we want when a mouse action is underway? 

Possible answers (brainstorming now, haven't thought much about the implications)
- The ability to trigger any key? In which case, what happens to the mouse action? 
- The ability to trigger only a specific key (e.g. “stop mouse action” or “press the same key again to stop”)
- The ability to trigger a subset of keys, e.g. mouse keys. Imagine you are doing continuous left clicks to navigate Windows Explorer. Now you want one right click and then to resume left clicking. This would be smooth and efficient if the 'continual' mouse action was paused to allow a single mouse action first. But not sure if this would be the natural UX in other situtions.

**Comment by AdamRoden** [comment]
@kmcnaught I did say it was just a start ;-) 

Not being able to stop a looping key is something that can happen in the current release. However it is handled needs to be failsafe.

Another brainstorm idea: Key triggers a popup with it's own specialized instance of a key fixation source with the purpose of always being on top of everything else and monitoring for fixation to end the key.

**Comment by JuliusSweetland** [comment]
I agree that 2 sounds like the best UX approach. Possible issues with
allowing key and point selection arise when the user is attempting mouse
actions over optikey itself. I think only showing specific keys to be
selected makes sense. A flexible approach might be to associate a set of
keys with the locked mouse action so only those are enabled when the mouse
key is selected - initially this might just be the same mouse key to allow
the user to toggle the mouse key state to locked and back to off.

**Comment by richardchui** [comment]
FWIW, I am using Irisbond Hiru's EasyClick software.  It has a separate button for Continuous Clicking which will remain in effect until another mode is selected.  And Continuous Clicking is only for the left mouse click which is good enough.

This is a feature that is extremely useful when the user is using gaze to control an AAC (picture cards) application.  With Continuous Clicking, he can keep on selecting different cards to communicate without having to revert back to the menu bar every time, a huge time saver.

**Comment by AdamRoden** [comment]
I think if we have a key in a continuous running state then we should process events from both key and point sources in parallel. I can imagine use cases where it would be desirable to stay in continuous click mode, but also be able to press a leter key to trigger some action in the game or program being used.

I would say make it so the running key can be stopped by it's own reactivation, by any other key that includes a KeyUp-All command, or if it is a Mouse type action, then by any other Mouse type actions.

**Comment by kmcnaught** [comment]
@JuliusSweetland 

I've now added a commit to the pending PR (565e2a2) which permits point and key selection to carry on independently along side each other. There is still some tidying up to do if we are happy with the direction. 

It was more complicated than expected as the triggers and callbacks for point and key selection were quite intertwined, with logic to tease out which trigger was relevant depending on current SelectionMode. This logic has all been teased apart now, which I think makes some of the code easier to read as you know whether you're dealing with point selection, key selection and gesture selection without needing to worry about lots on internal state. When events are created by the InputService, they have a `SelectionMode` variable so that consumers know where it came from. As I've already said, there is tidying up to be done but I believe this to be fully functional. 

With this commit, you can: 
- Select keys as per normal
- Select points as per normal 
- Perform gestures as per normal
- Select keys (currently any key) while a point action is also pending - I have been testing with key dwell < point dwell to make it easier to reason about. 
- As above but using dynamic keys for the point action and the key selection

Now need to play around with the discussed options for disabling certain keys, or not while a point action is pending.

**Comment by kmcnaught** [comment]
@AdamRoden is it intended that a `Loop Count="0"` key has a "locked down" appearance? This does not seem to be the case in either `origin/master` or Optikey Pro 3.2.2, although the key does perform as if it is toggled (by which I mean that I can cancel the loop by clicking it again). I am not sure if this is a regression, or something that wasn't ever implemented? It works in OK Game On but I don't remember if I added it?!

**Comment by AdamRoden** [comment]
İ don't believe İ was concerned with whether or not the lock icon displayed. Going off my memory, İ think there was some code related to it in OKGO that you changed that related to key down/locked down, so İ'm pretty sure your desired behavior was different. İ'm indifferent to appearance as long as it functions.

**Comment by kmcnaught** [comment]
An update, if anyone wants to play with the pending prototype implementation over at https://github.com/OptiKey/OptiKey/pull/874

You can lock the standard left click key and it will be on a loop. While locked down, you are still able to dwell on any other keys and they are functional. This is a proof of concept - there is some tidying up required of the refactor, and then decisions need to be made about the UX. 

I am going round in circles in what might be desirable wrt how much we restrict the simultaneous functionality of point and key dwelling, so am keen that we can experiment with the least restrictive implementation (i.e. the current behaviour) and then only restrict what we really need to. 

Some relevant thoughts - with no conclusion currently. 

### Questions

- Do you need to perform clicks on Optikey keyboard?
- Do you need to select keys while in a continuous clicking mode?

### Thoughts

- You may need to right click to access the context menu (unless we workaround this by providing an alternate method of access)

- Might you need to left click to get focus on Optikey? You may do this to use shortcuts for e.g. opening the management console, but if you are using eye gaze to open the management console you can just use the context menu instead. Can anyone think of another reason you might be left clicking on the Optikey keyboard?

- Giving focus to Optikey can actually be problematic, since you often are trying to send key input to other software. If you want to send a key press by dwelling on a key, you don't want your point dwell to click on the keyboard and steal focus just before your key dwell completes. 

- You obviously have to be able to select some keys in order to turn off the continuous modes. 

- I would assert that it is pretty important to be able to select arbitrary keys in the case of dynamic keyboards, where you may want to send a key or shortcut combo to an app in between clicking. For instance changing state while clicking in a game, or if you are using an app you may want to "click..click..click... Ctrl+O ... click..click..click"
 
- You may wish to be protected from accidentally selecting a key while you are trying to do a mouse click near the keyboard. Imagine Optikey is docked at the top of your screen and you want to click in your browser URL bar - you don't want the bottom Optikey keys fighting to be selected while you are trying to click. In this specific case, it would be simpler if you were only dwell selecting a mouse position and keys were disabled. 

- My guess is that in other eye gaze interfaces, it would not be expected that the interface stops reacting to dwell clicks while a continuous clicking action is underway BUT the click actions would not apply to the interface itself. I think the only reason that Optikey differs is that we require right click on the keyboard. 

- I am personally increasingly against the idea of disabling key dwell while a mouse action is active, but we might consider preventing mouse actions *on top of Optikey* if we can make this work in a reasonable way (make sure context menu is accessible, make sure we avoid weird edge behaviour when clicking *near* Optikey

**Comment by kmcnaught** [comment]
The approach agreed offline was:

- For a single mouse action (including optional magnification step) the behaviour should be as it is now, keys can not be selected and mouse actions can take place on Optikey if necessary. This gives people access to the context menu. 

- For a repeated mouse action, key selection and point selection are both functional, but the point selection can't take place over the Optikey keyboard. Points that are over the keyboard are just filtered out from the pipeline (so do not contribute to dwell progress).

**Comment by kmcnaught** [comment]
This is what the new behaviour looks like @JuliusSweetland, let me know if any of it doesn't look right or doesn't match what you were expecting. 

From my opinion the only thing potentially up for debate is whether the cursor popup should be visible over Optikey, but my vote is for yes since:
- that is simplest implementation wise
- hiding it might lead to weird behaviour near the edge of the Optikey keyboard
- it kind of feels symmetrical compared to keys being highlighted but not progressing when you are in single point action mode.

https://www.loom.com/share/c2defb56fa5646f8bee2a2a6212c0d77

**Comment by JuliusSweetland** [comment]
@kmcnaught Nailed it. I think showing the cursor has, as you say, symmetry with the unlocked point selection behaviour, as well as being a strong indicator that the user is still in locked clicking mode.

**Comment by kmcnaught** [comment]
I believe the implementation is now complete in the pending PR, though I will squash it before merging. There were some complications with how different features interacted with each other, I think I've found reasonable solutions but I'm mentioning them here for review. 

- When you are in a continuous point action mode and you choose another point action (e.g. left clicking continuously and then select single right click) it turns off any locked actions. If this proves to be a pain then we can revisit.

- If you are doing continuous point actions with magnification, you can't turn the action off whilst the magnifier is up. You will have to perform a click (on the magnified section or outside to cancel) before you can select keys again. This is a necessary limitation because (a) we need to be able to do a point selection on a part of the magnified popup which lies over the top of the keyboard and (b) we don't want to accidentally trigger keys that are hidden by the popup. 

- You _can_ do multi-key selection while also continuous clicking - they work independently. I can't imagine them being used like this but it was easiest (less logic) to just make sure neither breaks the other than to enforce any exclusivity. 

Basically my goal was to minimise the logic and state for each choice, let me know if anything sounds unreasonable.

**Comment by kmcnaught** [comment]
@JuliusSweetland

**Comment by JuliusSweetland** [comment]
I think those are all good decisions. Thanks Kirsty. We can catch up after
Christmas, but let me know if you want anything tested beforehand.

**Comment by kmcnaught** [comment]
It would be helpful for you to test the mouse locking (and try to break it) before we talk, but I don't need feedback before then.

**Comment by JuliusSweetland** [comment]
@kmcnaught Tested and looks great to me. :)

**Comment by kmcnaught** [comment]
@AdamRoden there is one remaining query as to how these lockable mouse actions should interact with your gestures. Previously I believe gestures only triggered during key selection mode (i.e. *not* while a point action is underway). Now that the point, key and gesture triggers have been separated out a bit, gestures can be triggered in any state. I suspect we want to add some limits here. 

We now have three selection states in Optikey:
`SinglePoint` - performing one point action. Keys are not selectable unless required to modify the point action (e.g. lock it down)
`ContinousPoints` - performing a locked point action on repeat until unlocked. Keys are useable in this state
`Keys` - only keys are selectable

I think it makes sense to disable gestures in the `SinglePoint` mode, since you might have a gesture which is "look at rectangle here" and you need to perform a mouse action in the same area. Do you agree? What about `ContinuousPoints`? I think the starting point would be to disable gestures here too, to avoid any situation where someone can't achieve what they want to, but do you have any arguments for having gestures during continuous clicking?

@JuliusSweetland we should resolve this before pre-release if possible (sorry I missed this todo)

**Comment by JuliusSweetland** [comment]
@kmcnaught understood. Let's see what Adam thinks.

**Comment by AdamRoden** [comment]
@JuliusSweetland @kmcnaught I've stated elsewhere that I'd like to remove the gestures feature. The 'Popup' key I added to dynamic keyboards provides users the ability to create off-screen keys, (which is probably the only real-world use case for gestures), and popup keys provide on-screen visual feedback, whereas with gestures there is no feedback. Also, the designer for dynamic keyboards is an evolutionary step forward from gestures. 

I vote to remove gestures.

**Comment by kmcnaught** [comment]
Okay, as a first step I've disabled gestures for SinglePoint and ContinuousPoints so we don't lose any functionality or gain any unexpected side effects. I don't currently understand enough about your new features to have an opinion on losing gestures. I don't use them myself or with anyone else but don't want to break workflows for existing users without a good migration plan. 

@JuliusSweetland there is a new PR with just this change now
